src/libasr/codegen/asr_to_llvm.cpp

```diff
@@ -4599,6 +4599,52 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
                 fmt.push_back("%lld");
                 llvm::Value* d = builder->CreatePtrToInt(tmp, getIntType(8, false));
                 args.push_back(d);
+            } else if (t->type == ASR::ttypeType::List) {
+                if(ASR::is_a<ASR::ListConstant_t>(*v)) {
+                    llvm::Value *open_bracket = builder->CreateGlobalStringPtr("[");
+                    fmt.push_back("%s");
+                    args.push_back(open_bracket);
+
+                    ASR::ListConstant_t* listC = ASR::down_cast<ASR::ListConstant_t>(v);
+                    ASR::ttype_t *arg_type;
+                    for(size_t j = 0; j < listC->n_args; ++j) {
+                        if (j == 0) {
+                            arg_type = expr_type(listC->m_args[j]); // current the arg_type in List is same
+                        } else {
+                            fmt.push_back("%s");
+                            llvm::Value *comma_space = builder->CreateGlobalStringPtr(", ");
+                            args.push_back(comma_space);
+                        }
+                        this->visit_expr_wrapper(listC->m_args[j], true);
+                        if (ASRUtils::is_integer(*arg_type)) {
+                            fmt.push_back("%d"); // not use "%lld" for ListConstant default i32
+                            args.push_back(tmp);
+                        }
+                        else if (ASRUtils::is_real(*arg_type)) {
+                            fmt.push_back("%f");
+                            args.push_back(builder->CreateFPExt(tmp, llvm::Type::getDoubleTy(context)));
+                        }
+                        else if (ASRUtils::is_character(*arg_type)) {
+                            llvm::Value *single_quote = builder->CreateGlobalStringPtr("'");
+                            fmt.push_back("%s");
+                            args.push_back(single_quote);
+                            fmt.push_back("%s");
+                            args.push_back(tmp);
+                            fmt.push_back("%s");
+                            args.push_back(single_quote);
+                        } else {
+                            throw LCompilersException("Printing ConstantList is only support for"
+                                "integer, float, char and string now.");
+                        }
+                    }
+                    fmt.push_back("%s");
+                    llvm::Value *close_bracket = builder->CreateGlobalStringPtr("]");
+                    args.push_back(close_bracket);
+                }
+                else {
+                    throw LCompilersException("Printing List is only support for "
+                        "ASR type ListConstant now.");
+                }
```

Comment:
I think we should do generic printing of list variables with the help of an ASR->ASR pass, so that all backend have the printing support at once. How I see printing it is as follows,

```python
print(l) # l is a list
```

should be converted to,

```python
print("[", end="")
for i in range(len(l)):
    if i < len(l) - 1:
        print(l[i], end=",")
    else:
        print(l[i], end="")
print("]")
```

Reply:
@certik What do you say?

Reply:
I thing it will be cool. A higher abstraction can save a lot of code, although I have few experience on asr to other backend now.